include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -714,10 +778,72 @@ namespace LinearAlgebra
 
 
     template <typename Number, typename MemorySpace>
-    void
-    SparseMatrix<Number, MemorySpace>::Tvmult(
-      Vector<Number, MemorySpace>       &dst,
-      const Vector<Number, MemorySpace> &src) const
+    template <typename InputVectorType>
+    std::enable_if_t<
+      !std::is_same_v<InputVectorType, Vector<Number, MemorySpace>>>
+    SparseMatrix<Number, MemorySpace>::vmult(InputVectorType       &dst,
+                                             const InputVectorType &src) const
+    {
+      Assert(&src != &dst, ExcSourceEqualsDestination());
+      Assert(matrix->isFillComplete(), ExcMatrixNotCompressed());
+
+      // get the size of the input vectors:
+      const size_type dst_local_size =
+        internal::end(dst) - internal::begin(dst);
+      const size_type src_local_size =
+        internal::end(src) - internal::begin(src);
+
+      // For the dst vector:
+      // create a Kokkos::View from the dst vector
+      typename VectorType::impl_scalar_type *raw_dst =
+        reinterpret_cast<typename VectorType::impl_scalar_type *>(
+          internal::begin(dst));
+      Kokkos::View<
+        typename VectorType::impl_scalar_type **,
+        Kokkos::LayoutLeft,
+        Kokkos::Device<typename MemorySpace::kokkos_space::execution_space,
+                       typename MemorySpace::kokkos_space>>
+        kokkos_view_dst(raw_dst, dst_local_size);
+
+      // get a Kokkos::DualView
+      auto mirror_view_dst = Kokkos::create_mirror_view(kokkos_view_dst);
+      typename VectorType::dual_view_type kokkos_dual_view_dst(kokkos_view_dst,
+                                                               mirror_view_dst);
+
+      // create the Tpetra::Vector
+      VectorType tpetra_dst(matrix->getRangeMap(), kokkos_dual_view_dst);
+
+      // For the src vector:
+      // create a Kokkos::View from the src vector
+      typename VectorType::impl_scalar_type *raw_src =
+        reinterpret_cast<typename VectorType::impl_scalar_type *>(
```

Comment:
Why do you need a `reinterpret_cast` here?

Reply:
It seems like the `Tpetra::Vector` expects this type. 
Since this is just a complicated alias for `Number`, I can give it a try and remove this cast.

Reply:
I would be very suspicious of any `reinterpret_cast` we are doing in these classes since we don't really have a buffer that we would use with placement new. Whenever we need a different type we should copy values instead.

Reply:
I agree with this. `x.begin()` does not always return a pointer. It could just be an iterator. `reinterpret_cast` will in such cases not result in anything useful, with no error message. I think there ought to be a better way.

What *are* the types you're casting here?

Reply:
We do not need the `reinterpret_cast` here. But at another point, we need a `const_cast`.
I specialized the function so we know which vector type we are working on, so we can ensure that the cast is working correctly.